lib/server.go

```diff
@@ -1728,16 +1790,36 @@ func (srv *Server) _addNewTxn(
 	}
 
 	srv.blockchain.ChainLock.RLock()
-	newlyAcceptedTxns, err := srv.mempool.ProcessTransaction(
-		txn, true /*allowUnconnectedTxn*/, rateLimit, peerID, verifySignatures)
-	srv.blockchain.ChainLock.RUnlock()
-	if err != nil {
-		return nil, errors.Wrapf(err, "Server._handleTransaction: Problem adding transaction to mempool: ")
+	defer srv.blockchain.ChainLock.RUnlock()
+
+	// Only attempt to add the transaction to the PoW mempool if we're on the
+	// PoW protocol. If we're on the PoW protocol, then we use the PoW mempool's,
+	// txn validity checks to signal whether the txn has been added or not. The PoW
+	// mempool has stricter txn validity checks than the PoW mempool, so this works
+	// out conveniently, as it allows us to always add a txn to the PoS mempool.
+	blockHeight := srv.blockchain.BlockTip().Height
+	if blockHeight < srv.blockchain.params.ForkHeights.ProofOfStake2ConsensusCutoverBlockHeight {
+		_, err := srv.mempool.ProcessTransaction(
+			txn, true /*allowUnconnectedTxn*/, rateLimit, peerID, verifySignatures)
+		if err != nil {
+			return nil, errors.Wrapf(err, "Server._addNewTxn: Problem adding transaction to mempool: ")
+		}
+
+		// At this point, we know the transaction has been run through the mempool.
+		// Now wait for an update of the ReadOnlyUtxoView so we don't break anything.
+		srv.mempool.BlockUntilReadOnlyViewRegenerated()
```

Comment:
I think there could be a problem if you hold the the ChainLock.RLock() while you wait for this thing. Maybe not. It does take like a whole second or two, though, and I think kindof bad to hold the RLock for that long maybe idk. Think about it.

Reply:
Hmm, good call. We really only need to hold the `ChainLock.RLock()` while reading the blockTip's height. I've updated the code to reflect that